x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -115,96 +165,127 @@ private void loadModel(String modelId) {
+                }
+            }
+            removedModels = Sets.difference(referencedModelsBeforeClusterState, allReferencedModelKeys);
 
-                // Remove all that are currently being loaded
-                allReferencedModelKeys.removeAll(loadingListeners.keySet());
+            // Remove all cached models that are not referenced by any processors
+            removedModels.forEach(localModelCache::invalidate);
+            // Remove the models that are no longer referenced
+            referencedModels.removeAll(removedModels);
+            shouldNotAudit.removeAll(removedModels);
 
-                // Remove all that are fully loaded, will attempt empty model loading again
-                loadedModels.forEach((id, optionalModel) -> {
-                    if (optionalModel.isSuccess()) {
-                        allReferencedModelKeys.remove(id);
-                    }
-                });
-                // Populate loadingListeners key so we know that we are currently loading the model
-                for (String modelId : allReferencedModelKeys) {
-                    loadingListeners.put(modelId, new ArrayDeque<>());
-                }
-                if (loadedModelBeforeClusterState != null && loadingModelBeforeClusterState != null) {
-                    if (loadingListeners.keySet().equals(loadingModelBeforeClusterState) == false) {
-                        logger.trace("cluster state event changed loading models: before {} after {}", loadingModelBeforeClusterState,
-                            loadingListeners.keySet());
-                    }
-                    if (loadedModels.keySet().equals(loadedModelBeforeClusterState) == false) {
-                        logger.trace("cluster state event changed loaded models: before {} after {}", loadedModelBeforeClusterState,
-                            loadedModels.keySet());
-                    }
-                }
+            // Remove all that are still referenced, i.e. the intersection of allReferencedModelKeys and referencedModels
+            allReferencedModelKeys.removeAll(referencedModels);
+            referencedModels.addAll(allReferencedModelKeys);
 
+            // Populate loadingListeners key so we know that we are currently loading the model
+            for (String modelId : allReferencedModelKeys) {
+                loadingListeners.put(modelId, new ArrayDeque<>());
             }
-            for (Tuple<String, List<ActionListener<Model>>> modelAndListeners : drainWithFailure) {
-                final String msg = new ParameterizedMessage(
-                    "Cancelling load of model [{}] as it is no longer referenced by a pipeline",
-                    modelAndListeners.v1()).getFormat();
-                for (ActionListener<Model> listener : modelAndListeners.v2()) {
-                    listener.onFailure(new ElasticsearchException(msg));
-                }
+        }
```

Comment:
nit: would be good for reading to mark the end of a synchronized block as comment e.g. `} // synchronized(...)`